contracts/cw-names-registry/src/msg.rs

```diff
@@ -0,0 +1,62 @@
+use cosmwasm_std::{Addr, Uint128};
+use cw20::Cw20ReceiveMsg;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct InstantiateMsg {
+    pub admin: String,
+    pub payment_token_address: String,
+    pub payment_amount: Uint128,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum ExecuteMsg {
+    Receive(Cw20ReceiveMsg), // Receive payment to register a name
+    UpdateConfig {
+        new_payment_token_address: Option<String>,
+        new_admin: Option<String>,
+        new_payment_amount: Option<Uint128>,
+    },
+    /// Reserve a name so it cannot be taken for later use
+    Reserve {
+        name: String,
+    },
+    /// Transfer a reserved name to a DAO
+    TransferReservation {
+        name: String,
+        dao: String,
+    },
+    Revoke {
+        name: String,
+    },
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum ReceiveMsg {
+    /// DAO can register a name by paying, we assume DAO is the sender
+    Register { name: String },
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum QueryMsg {
+    Config {},
+    LookUpDao { dao: String },
+    LookUpName { name: String },
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct LookUpDaoResponse {
+    pub name: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct LookUpNameResponse {
```

Comment:
I think there may be two routes in here: (1) something I'd call `LookUpDaoByName`, which you might call when you're searching for a DAO, and (2) something I'd call `IsNameAvailableToRegister`.  I'd expect the former to return an Option`<Addr>` and the latter to return a `bool`. Is there a reason not to separate them out?

Now, maybe the answer is, "this way, you only need to do one query to see which DAO has a name + whether that name is available," which, fine, but will users ever need to know these things at the same time? I think people calling this contract (for now, the proposal template) will mostly want to know whether a DAO is available - and we can give them a really simple route to call to find out. 

Not going to bikeshed on this point though - it's almost certainly fine either way.